src/language-js/printer-estree.js

```diff
@@ -2324,7 +2324,9 @@ function printPathNoParens(path, options, print, args) {
     case "TSTupleType":
     case "TupleTypeAnnotation": {
       const typesField = n.type === "TSTupleType" ? "elementTypes" : "types";
+      // TODO: Remove after not-last rest tuple type elements by typescript-estree
       const hasRest =
+        options.parser !== "babel-ts" &&
```

Comment:
This is temporary, right?

Reply:
Yes.

Reply:
If I understand correctly, we can't print trailing comma for `typescript` parser, but we can in TS4?

Reply:
Sorry oops, I misunderstood. TS 3.9.x doesn't support non-rest elements in tuple types but can parse! We can remove `hasRest`.
I'll fix this PR. After that, please review again.

Reply:
Sorry, I misunderstood. Current typescript-estree can parse not-last rest elements... I'll fix this PR.